File: src/server/controllers/maps-generate.ts
import { randomUUID } from 'crypto';
import express from 'express';
import { z } from 'zod';

import {
  buildMapExportConfig,
  runMapGenerateConversion,
} from '@/lib/converter/map-exporter/run-map-generate';
import { getDefaultConfig } from '@/lib/global-config';
import { Vector2 } from '@/lib/math/common';
import { buildADTExportOptions } from '@/lib/wow/export/adt/map-export-utils';
import { computeStepsPerTile } from '@/lib/wow/export/export-progress';
import { wowExportClient } from '@/lib/wowexport-client/wowexport-client';
import {
  advanceMapGenerateProgress,
  clearMapGenerateProgress,
  getMapGenerateProgress,
  initMapGenerateProgress,
  setMapGeneratePhase,
  syncAdtProgress,
  toProgressPercent,
  updateMapGenerateTotalSteps,
} from '@/server/map-generate-progress';
import { Job, JobQueue } from '@/server/utils/job-queue';

import type { MapExportTileFailure, MapExportTileSuccess } from './maps-export';

const tileSchema = z.object({
  x: z.number().int().min(0).max(63),
  y: z.number().int().min(0).max(63),
});

const qualitySchema = z.union([
  z.literal(0),
  z.literal(512),
  z.literal(1024),
  z.literal(2048),
  z.literal(4096),
  z.literal(8192),
  z.literal(16384),
]);

export const generateWc3BodySchema = z.object({
  tiles: z.array(tileSchema).min(1),
  quality: qualitySchema,
  mapSaveName: z.string().trim().min(1).max(128)
    .regex(/^[a-zA-Z0-9_.-]+(\.w3x)?$/i),
  clampLower: z.number().min(0).max(1),
  clampUpper: z.number().min(0).max(1),
  mapAngleDeg: z.number(),
  freshExport: z.boolean(),
  creatures: z.object({
    enable: z.boolean(),
    allAreDoodads: z.boolean(),
    scaleUp: z.number().positive(),
  }),
}).strict().refine((d) => d.clampUpper >= d.clampLower, {
  message: 'clampUpper must be >= clampLower',
  path: ['clampUpper'],
});

export type GenerateWc3Body = z.infer<typeof generateWc3BodySchema>;

export interface MapGenerateJobRequest {
  mapDir: string;
  mapID: number;
  body: GenerateWc3Body;
  orderedTiles: readonly { x: number; y: number }[];
  tileBounds: { min: Vector2; max: Vector2 };
}

export interface MapGenerateJobResult {
  id: 'WC3_MAP_GENERATE_SUMMARY';
  map: string;
  mapID: number;
  mapSaveName: string;
  outputDir: string;
  quality: number;
  total: number;
  stepsPerTile: number;
  totalSteps: number;
  succeeded: MapExportTileSuccess[];
  failed: MapExportTileFailure[];
}

export interface MapGenerateProgress {
  completedSteps: number;
  totalSteps: number;
  phase: 'adt' | 'convert';
  tileIndex?: number;
  tileCount?: number;
  stepsPerTile?: number;
  currentTile?: { x: number; y: number };
  taskName?: string;
  creatureCompleted?: number;
  creatureTotal?: number;
  percent: number;
}

export interface MapGenerateJobStatus {
  id: string;
  status: 'pending' | 'processing' | 'done' | 'failed';
  position?: number;
  mapSaveName?: string;
  mapDir?: string;
  queuePending?: number;
  result?: MapGenerateJobResult;
  error?: string;
  submittedAt: number;
  startedAt?: number;
  finishedAt?: number;
  progress?: MapGenerateProgress;
}

type MapGenerateJob = Job<MapGenerateJobRequest, MapGenerateJobResult>;

const INITIAL_CONVERT_STEPS = 3;

function tileBoundsFromTiles(tiles: readonly { x: number; y: number }[]): { min: Vector2; max: Vector2 } {
  const xs = tiles.map((t) => t.x);
  const ys = tiles.map((t) => t.y);
  return {
    min: [Math.min(...xs), Math.min(...ys)],
    max: [Math.max(...xs), Math.max(...ys)],
  };
}

async function pollAdtProgress(
  jobId: string,
  adtTotalSteps: number,
  until: () => boolean,
): Promise<void> {
  while (!until()) {
    const snap = await wowExportClient.getExportProgress(jobId);
    if (snap) {
      syncAdtProgress(jobId, {
        completedSteps: Math.min(snap.completedSteps, adtTotalSteps),
        tileIndex: snap.tileIndex,
        tileCount: snap.tileCount,
        stepsPerTile: snap.stepsPerTile,
        currentTile: snap.currentTile,
        taskName: snap.taskName,
      });
    }
    await new Promise((r) => { setTimeout(r, 250); });
  }
}

const mapGenerateQueue = new JobQueue<MapGenerateJobRequest, MapGenerateJobResult>(
  {
    concurrency: 1,
    maxPendingJobs: 10,
    jobTTL: 10 * 60 * 1000,
    jobTimeout: 2 * 60 * 60 * 1000,
  },
  async (job) => {
    const {
      mapDir, mapID, body, orderedTiles, tileBounds,
    } = job.request;
    const progressKey = job.id;

    const exportOptions = buildADTExportOptions(undefined, {
      mapsIncludeM2: true,
      mapsIncludeWMO: true,
      mapsIncludeWMOSets: true,
      mapsIncludeGameObjects: true,
      mapsIncludeLiquid: true,
      mapsIncludeFoliage: true,
      mapsIncludeHoles: true,
    });
    const stepsPerTile = computeStepsPerTile(body.quality, exportOptions);
    const tileCount = orderedTiles.length;
    const adtTotalSteps = tileCount * stepsPerTile;

    initMapGenerateProgress(progressKey, {
      adtTotalSteps,
      convertSteps: INITIAL_CONVERT_STEPS,
      tileCount,
      stepsPerTile,
    });

    await wowExportClient.waitUntilReady();

    const succeeded: MapExportTileSuccess[] = [];
    const failed: MapExportTileFailure[] = [];

    for (let tileIndex = 0; tileIndex < orderedTiles.length; tileIndex++) {
      const { x: tileX, y: tileY } = orderedTiles[tileIndex];
      let pollDone = false;
      const pollPromise = pollAdtProgress(progressKey, adtTotalSteps, () => pollDone);

      try {
        const result = await wowExportClient.exportADT({
          mapID,
          mapDir,
          tileX,
          tileY,
          quality: body.quality,
          includeM2: true,
          includeWMO: true,
          includeWMOSets: true,
          includeGameObjects: true,
          includeLiquid: true,
          includeFoliage: true,
          includeHoles: true,
          progressKey,
          tileIndex,
          tileCount,
          stepsPerTile,
        });
        succeeded.push({ tileX, tileY, result });
      } catch (e) {
        failed.push({ tileX, tileY, error: e instanceof Error ? e.message : String(e) });
      } finally {
        pollDone = true;
        await pollPromise;
        syncAdtProgress(progressKey, {
          completedSteps: Math.min((tileIndex + 1) * stepsPerTile, adtTotalSteps),
          tileIndex,
          tileCount,
          stepsPerTile,
          currentTile: { x: tileX, y: tileY },
          taskName: tileIndex + 1 >= tileCount ? 'Tiles exported' : 'Exporting tiles',
        });
      }
    }

    await wowExportClient.finalizeExportProgress(progressKey);

    if (failed.length === orderedTiles.length) {
      clearMapGenerateProgress(progressKey);
      throw new Error(failed[0]?.error ?? 'All tiles failed to export');
    }

    setMapGeneratePhase(progressKey, 'convert', 'Converting to WC3 map');
    syncAdtProgress(progressKey, {
      completedSteps: adtTotalSteps,
      tileIndex: tileCount - 1,
      tileCount,
      stepsPerTile,
      taskName: 'Tiles exported',
    });

    const config = await getDefaultConfig();
    const mapExportConfig = buildMapExportConfig({
      mapId: mapID,
      wowExportFolder: mapDir.toLowerCase(),
      min: tileBounds.min,
      max: tileBounds.max,
      mapAngleDeg: body.mapAngleDeg,
      clampLower: body.clampLower,
      clampUpper: body.clampUpper,
      creatures: body.creatures,
    });

    let convertSteps = INITIAL_CONVERT_STEPS;
    const conversion = await runMapGenerateConversion({
      config,
      mapExportConfig,
      mapSaveName: body.mapSaveName,
      freshExport: body.freshExport,
      creatureScaleUp: body.creatures.scaleUp,
      onConvertStepsKnown: (steps) => {
        convertSteps = steps;
        updateMapGenerateTotalSteps(progressKey, convertSteps, adtTotalSteps);
      },
      onProgress: (convertCompleted, taskName, creatureProgress) => {
        advanceMapGenerateProgress(
          progressKey,
          adtTotalSteps,
          convertCompleted,
          taskName,
          creatureProgress,
        );
      },
    });

    const totalSteps = adtTotalSteps + conversion.convertSteps;
    advanceMapGenerateProgress(progressKey, adtTotalSteps, conversion.convertSteps, 'Complete');

    return {
      id: 'WC3_MAP_GENERATE_SUMMARY',
      map: mapDir,
      mapID,
      mapSaveName: conversion.mapSaveName,
      outputDir: conversion.outputDir,
      quality: body.quality,
      total: orderedTiles.length,
      stepsPerTile,
      totalSteps,
      succeeded,
      failed,
    };
  },
);

function buildProgressStatus(jobId: string): MapGenerateProgress | undefined {
  const snap = getMapGenerateProgress(jobId);
  if (!snap) return undefined;
  return {
    completedSteps: snap.completedSteps,
    totalSteps: snap.totalSteps,
    phase: snap.phase,
    tileIndex: snap.tileIndex,
    tileCount: snap.tileCount,
    stepsPerTile: snap.stepsPerTile,
    currentTile: snap.currentTile,
    taskName: snap.taskName,
    creatureCompleted: snap.creatureCompleted,
    creatureTotal: snap.creatureTotal,
    percent: toProgressPercent(snap),
  };
}

function enrichJobMeta(status: MapGenerateJobStatus, job: MapGenerateJob): void {
  status.mapSaveName = job.request.body.mapSaveName;
  status.mapDir = job.request.mapDir;
  status.queuePending = mapGenerateQueue.getQueueSnapshot().pendingCount;
}

async function buildJobStatus(jobId: string): Promise<MapGenerateJobStatus | undefined> {
  const base = mapGenerateQueue.getJobStatus(jobId);
  const job = mapGenerateQueue.getJob(jobId);
  if (!base || !job) return undefined;

  const status: MapGenerateJobStatus = {
    id: base.id,
    status: base.status,
    position: base.position,
    result: base.result,
    error: base.error,
    submittedAt: base.submittedAt,
    startedAt: base.startedAt,
    finishedAt: base.finishedAt,
  };
  enrichJobMeta(status, job);

  if (base.status === 'processing') {
    if (getMapGenerateProgress(jobId)) {
      status.progress = buildProgressStatus(jobId);
    } else {
      const adtSnap = await wowExportClient.getExportProgress(jobId);
      if (adtSnap) {
        status.progress = {
          completedSteps: adtSnap.completedSteps,
          totalSteps: adtSnap.totalSteps,
          phase: 'adt',
          tileIndex: adtSnap.tileIndex,
          tileCount: adtSnap.tileCount,
          stepsPerTile: adtSnap.stepsPerTile,
          currentTile: adtSnap.currentTile,
          taskName: adtSnap.taskName,
          percent: adtSnap.totalSteps > 0
            ? Math.min(100, Math.round((adtSnap.completedSteps / adtSnap.totalSteps) * 100))
            : 0,
        };
      }
    }
  } else if (base.status === 'done' && base.result) {
    status.progress = {
      completedSteps: base.result.totalSteps,
      totalSteps: base.result.totalSteps,
      phase: 'convert',
      taskName: 'Complete',
      percent: 100,
    };
    clearMapGenerateProgress(jobId);
  } else if (base.status === 'failed') {
    clearMapGenerateProgress(jobId);
  }

  return status;
}

export function registerMapGenerateRoutes(
  router: express.Router,
  resolveMap: (key: string) => { id: number; dir: string } | undefined,
): void {
  router.post('/maps/:map/generate-wc3', (req, res) => {
    try {
      const key = String(req.params.map).toLowerCase();
      const entry = resolveMap(key);
      if (!entry) return res.status(404).json({ error: 'Unknown map' });

      const parsed = generateWc3BodySchema.safeParse(req.body);
      if (!parsed.success) {
        return res.status(400).json({ error: 'Invalid request body', issues: z.treeifyError(parsed.error) });
      }

      const unique = new Map<string, { x: number; y: number }>();
      for (const t of parsed.data.tiles) unique.set(`${t.x},${t.y}`, t);
      const orderedTiles = Array.from(unique.values()).sort((a, b) => (a.y - b.y) || (a.x - b.x));

      const job: MapGenerateJob = {
        id: randomUUID(),
        request: {
          mapDir: entry.dir,
          mapID: entry.id,
          body: parsed.data,
          orderedTiles,
          tileBounds: tileBoundsFromTiles(orderedTiles),
        },
        status: 'pending',
        submittedAt: Date.now(),
        noTimeout: true,
      };

      mapGenerateQueue.addJob(job);
      console.log('generate-wc3 queued', key, job.id, `${orderedTiles.length} tiles → ${parsed.data.mapSaveName}`);

      setImmediate(() => {
        void buildJobStatus(job.id).then((status) => {
          res.json(status);
        });
      });
      return undefined;
    } catch (e) {
      return res.status(500).json({ error: e instanceof Error ? e.message : String(e) });
    }
  });

  router.get('/maps/generate-wc3/status/:jobId', (req, res) => {
    void (async () => {
      const status = await buildJobStatus(req.params.jobId);
      if (!status) {
        res.status(404).json({ error: 'Generate job not found' });
        return;
      }
      res.json(status);
    })();
  });

  router.get('/maps/generate-wc3/active', (_req, res) => {
    void (async () => {
      const ids = mapGenerateQueue.listActiveJobIds();
      const jobs = await Promise.all(ids.map((id) => buildJobStatus(id)));
      res.json(jobs.filter((j): j is MapGenerateJobStatus => j != null));
    })();
  });
}

export { tileBoundsFromTiles };

File: golang/internal/server/api/maps_generate.go
package api

import (
	"context"
	"log"
	"net/http"
	"runtime"
	"runtime/debug"
	"sort"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/pqhuy98/wow-converter/internal/config"
	"github.com/pqhuy98/wow-converter/internal/converter/common"
	"github.com/pqhuy98/wow-converter/internal/converter/mapexporter"
	"github.com/pqhuy98/wow-converter/internal/math"
	"github.com/pqhuy98/wow-converter/internal/server/util"
	exportadt "github.com/pqhuy98/wow-converter/internal/wow/export/adt"
	"github.com/pqhuy98/wow-converter/internal/wow/casc"
)

type generateWc3Body struct {
	Tiles        []exportAdtTile `json:"tiles"`
	Quality      int             `json:"quality"`
	MapSaveName  string          `json:"mapSaveName"`
	ClampLower       float64 `json:"clampLower"`
	ClampUpper       float64 `json:"clampUpper"`
	AutoClampPercent bool    `json:"autoClampPercent"`
	MapAngleDeg      float64 `json:"mapAngleDeg"`
	UnitScale    float64         `json:"unitScale"`
	IncludeBuildingInteriors *bool           `json:"includeBuildingInteriors"`
	FreshExport              bool            `json:"freshExport"`
	Creatures    struct {
		Enable        bool `json:"enable"`
		AllAreDoodads bool `json:"allAreDoodads"`
	} `json:"creatures"`
}

type mapGenerateJobRequest struct {
	MapDir       string
	MapID        int
	Body         generateWc3Body
	OrderedTiles []exportAdtTile
	TileBounds   tileBounds
}

type tileBounds struct {
	Min [2]int `json:"min"`
	Max [2]int `json:"max"`
}

type mapGenerateJobResult struct {
	ID          string                 `json:"id"`
	Map         string                 `json:"map"`
	MapID       int                    `json:"mapID"`
	MapSaveName string                 `json:"mapSaveName"`
	OutputDir   string                 `json:"outputDir"`
	Quality     int                    `json:"quality"`
	Total       int                    `json:"total"`
	StepsPerTile int                   `json:"stepsPerTile"`
	TotalSteps  int                    `json:"totalSteps"`
	Succeeded   []mapExportTileSuccess `json:"succeeded"`
	Failed      []mapExportTileFailure `json:"failed"`
}

type mapGenerateProgressView struct {
	CompletedSteps    int        `json:"completedSteps"`
	TotalSteps        int        `json:"totalSteps"`
	Phase             string     `json:"phase"`
	TileIndex         *int       `json:"tileIndex,omitempty"`
	TileCount         *int       `json:"tileCount,omitempty"`
	StepsPerTile      *int       `json:"stepsPerTile,omitempty"`
	CurrentTile       *tileCoord `json:"currentTile,omitempty"`
	TaskName          string     `json:"taskName,omitempty"`
	CreatureCompleted *int       `json:"creatureCompleted,omitempty"`
	CreatureTotal     *int       `json:"creatureTotal,omitempty"`
	Percent           int        `json:"percent"`
}

type mapGenerateJobStatus struct {
	ID           string                   `json:"id"`
	Status       util.JobStatus           `json:"status"`
	Position     *int                     `json:"position,omitempty"`
	MapSaveName  string                   `json:"mapSaveName,omitempty"`
	MapDir       string                   `json:"mapDir,omitempty"`
	QueuePending *int                     `json:"queuePending,omitempty"`
	Result       *mapGenerateJobResult    `json:"result,omitempty"`
	Error        string                   `json:"error,omitempty"`
	SubmittedAt  int64                    `json:"submittedAt"`
	StartedAt    *int64                   `json:"startedAt,omitempty"`
	FinishedAt   *int64                   `json:"finishedAt,omitempty"`
	Progress     *mapGenerateProgressView `json:"progress,omitempty"`
}

const initialConvertSteps = 3

func registerMapGenerateRoutes(r Router, d *Deps) {
	queue := util.NewJobQueue(mapGenerateQueueConfig(), func(job *util.Job[mapGenerateJobRequest, mapGenerateJobResult]) (mapGenerateJobResult, error) {
		return runMapGenerateJob(context.Background(), d, job)
	})

	r.Post("/maps/{map}/generate-wc3", func(w http.ResponseWriter, req *http.Request) {
		if err := assertDesktopOnly(d.Config.IsSharedHosting); err != nil {
			sendError(w, http.StatusForbidden, err.Error())
			return
		}
		mapID, mapDir, ok := resolveMap(chi.URLParam(req, "map"))
		if !ok {
			sendError(w, http.StatusNotFound, "Unknown map")
			return
		}
		var body generateWc3Body
		if err := readJSONBody(req, &body); err != nil || len(body.Tiles) == 0 || body.MapSaveName == "" {
			sendError(w, http.StatusBadRequest, "Invalid request body")
			return
		}
		body.MapSaveName = mapexporter.NormalizeMapSaveName(body.MapSaveName)
		if body.MapSaveName == "" {
			sendError(w, http.StatusBadRequest, "Invalid map save name")
			return
		}
		if body.ClampUpper < body.ClampLower {
			sendError(w, http.StatusBadRequest, "clampUpper must be >= clampLower")
			return
		}

		ordered := dedupeTiles(body.Tiles)
		job := &util.Job[mapGenerateJobRequest, mapGenerateJobResult]{
			ID: newJobID(),
			Request: mapGenerateJobRequest{
				MapDir: mapDir, MapID: mapID, Body: body,
				OrderedTiles: ordered, TileBounds: tileBoundsFromTiles(ordered),
			},
			Status: util.JobPending, SubmittedAt: time.Now().UnixMilli(), NoTimeout: true,
		}
		queue.AddJob(job)
		log.Printf("generate-wc3 queued %s %s %d tiles → %s", mapDir, job.ID, len(ordered), body.MapSaveName)
		status := buildMapGenerateStatus(d, queue, job.ID)
		sendJSON(w, http.StatusOK, status)
	})

	r.Get("/maps/generate-wc3/status/{jobId}", func(w http.ResponseWriter, req *http.Request) {
		if err := assertDesktopOnly(d.Config.IsSharedHosting); err != nil {
			sendError(w, http.StatusForbidden, err.Error())
			return
		}
		status := buildMapGenerateStatus(d, queue, chi.URLParam(req, "jobId"))
		if status == nil {
			sendError(w, http.StatusNotFound, "Generate job not found")
			return
		}
		sendJSON(w, http.StatusOK, status)
	})

	r.Get("/maps/generate-wc3/active", func(w http.ResponseWriter, _ *http.Request) {
		if err := assertDesktopOnly(d.Config.IsSharedHosting); err != nil {
			sendError(w, http.StatusForbidden, err.Error())
			return
		}
		ids := queue.ListActiveJobIDs()
		out := make([]*mapGenerateJobStatus, 0, len(ids))
		for _, id := range ids {
			if status := buildMapGenerateStatus(d, queue, id); status != nil {
				out = append(out, status)
			}
		}
		sendJSON(w, http.StatusOK, out)
	})
}

func mapGenerateQueueConfig() util.QueueConfig[mapGenerateJobRequest, mapGenerateJobResult] {
	return util.QueueConfig[mapGenerateJobRequest, mapGenerateJobResult]{
		Concurrency: 1, MaxPendingJobs: 10, JobTTL: 10 * time.Minute, JobTimeout: 2 * time.Hour,
	}
}

func runMapGenerateJob(ctx context.Context, d *Deps, job *util.Job[mapGenerateJobRequest, mapGenerateJobResult]) (mapGenerateJobResult, error) {
	req := job.Request
	progressKey := job.ID
	includeInteriors := includeBuildingInteriorsEnabled(req.Body.IncludeBuildingInteriors)
	stepsPerTile := 1
	tileCount := len(req.OrderedTiles)
	adtTotalSteps := tileCount * stepsPerTile

	initMapGenerateProgress(progressKey, adtTotalSteps, initialConvertSteps, tileCount, stepsPerTile)

	mapExportCfg := mapexporter.BuildMapExportConfig(struct {
		MapID           int
		WowExportFolder string
		Min             math.Vector2
		Max             math.Vector2
		MapAngleDeg     float64
		ClampLower      float64
		ClampUpper      float64
		UnitScale       float64
		CreaturesEnable bool
		AllAreDoodads   bool
	}{
		MapID:           req.MapID,
		WowExportFolder: req.MapDir,
		Min:             math.Vector2{float64(req.TileBounds.Min[0]), float64(req.TileBounds.Min[1])},
		Max:             math.Vector2{float64(req.TileBounds.Max[0]), float64(req.TileBounds.Max[1])},
		MapAngleDeg:     req.Body.MapAngleDeg,
		ClampLower:      req.Body.ClampLower,
		ClampUpper:      req.Body.ClampUpper,
		UnitScale:       req.Body.UnitScale,
		CreaturesEnable: req.Body.Creatures.Enable,
		AllAreDoodads:   req.Body.Creatures.AllAreDoodads,
	})

	registry := common.NewTileRegistry()
	tileCoords := make([]mapexporter.TileCoord, len(req.OrderedTiles))
	for i, t := range req.OrderedTiles {
		tileCoords[i] = mapexporter.TileCoord{X: t.X, Y: t.Y}
	}
	if err := mapexporter.LoadADTTilesForConversion(ctx, d.Client, d.ExportAssetDir(), mapExportCfg, req.Body.Quality, tileCoords, includeInteriors, registry, func(completed, total int, tile mapexporter.TileCoord) {
		syncAdtProgress(progressKey, minInt(completed*stepsPerTile, adtTotalSteps), completed-1, tileCount, stepsPerTile,
			&tileCoord{X: tile.X, Y: tile.Y}, "Loading tiles")
	}); err != nil {
		clearMapGenerateProgress(progressKey)
		return mapGenerateJobResult{}, err
	}
	exportadt.ReleaseAdtExportBatchMemory()
	runtime.GC()
	debug.FreeOSMemory()

	succeeded := make([]mapExportTileSuccess, 0, len(req.OrderedTiles))
	for _, t := range req.OrderedTiles {
		succeeded = append(succeeded, mapExportTileSuccess{
			TileX: t.X, TileY: t.Y,
			Result: casc.ADTExportResult{ExportType: "ADT_DIRECT", MapID: req.MapID, MapDir: req.MapDir, TileX: t.X, TileY: t.Y},
		})
	}
	failed := make([]mapExportTileFailure, 0)

	setMapGeneratePhase(progressKey, phaseConvert, "Converting to WC3 map")
	syncAdtProgress(progressKey, adtTotalSteps, tileCount-1, tileCount, stepsPerTile, nil, "Tiles loaded")

	cfg := config.DefaultConfig()
	cfg.ExportAssetDir = d.ExportAssetDir()
	convertSteps := initialConvertSteps
	conversion, err := mapexporter.RunMapGenerateConversion(ctx, mapexporter.MapGenerateConversionOptions{
		Config:                   cfg,
		MapExportConfig:          &mapExportCfg,
		MapSaveName:              req.Body.MapSaveName,
		FreshExport:              req.Body.FreshExport,
		IncludeBuildingInteriors: includeInteriors,
		AutoClampPercent:         req.Body.AutoClampPercent,
		UnitScale:        req.Body.UnitScale,
		WowClient:        d.Client,
		TileRegistry:     registry,
		TileQuality:      req.Body.Quality,
		OnConvertStepsKnown: func(steps int) {
			convertSteps = steps
			updateMapGenerateTotalSteps(progressKey, convertSteps, adtTotalSteps)
		},
		OnProgress: func(convertCompleted int, taskName string, creature *mapexporter.CreatureProgress) {
			var cc, ct *int
			if creature != nil {
				cc, ct = &creature.Completed, &creature.Total
			}
			advanceMapGenerateProgress(progressKey, adtTotalSteps, convertCompleted, taskName, cc, ct)
		},
	})
	if err != nil {
		clearMapGenerateProgress(progressKey)
		return mapGenerateJobResult{}, err
	}

	totalSteps := adtTotalSteps + conversion.ConvertSteps
	advanceMapGenerateProgress(progressKey, adtTotalSteps, conversion.ConvertSteps, "Complete", nil, nil)

	return mapGenerateJobResult{
		ID:           "WC3_MAP_GENERATE_SUMMARY",
		Map:          req.MapDir,
		MapID:        req.MapID,
		MapSaveName:  conversion.MapSaveName,
		OutputDir:    conversion.OutputDir,
		Quality:      req.Body.Quality,
		Total:        tileCount,
		StepsPerTile: stepsPerTile,
		TotalSteps:   totalSteps,
		Succeeded:    succeeded,
		Failed:       failed,
	}, nil
}

func boolPtr(v bool) *bool { return &v }

func tileBoundsFromTiles(tiles []exportAdtTile) tileBounds {
	xs := make([]int, len(tiles))
	ys := make([]int, len(tiles))
	for i, t := range tiles {
		xs[i], ys[i] = t.X, t.Y
	}
	sort.Ints(xs)
	sort.Ints(ys)
	return tileBounds{Min: [2]int{xs[0], ys[0]}, Max: [2]int{xs[len(xs)-1], ys[len(ys)-1]}}
}

func buildMapGenerateStatus(d *Deps, queue *util.JobQueue[mapGenerateJobRequest, mapGenerateJobResult], jobID string) *mapGenerateJobStatus {
	base := queue.GetJobStatus(jobID)
	job := queue.GetJob(jobID)
	if base == nil || job == nil {
		return nil
	}
	pending, _ := queue.GetQueueSnapshot()
	status := &mapGenerateJobStatus{
		ID: base.ID, Status: base.Status, Position: base.Position,
		MapSaveName: job.Request.Body.MapSaveName, MapDir: job.Request.MapDir,
		QueuePending: &pending, Error: base.Error, SubmittedAt: base.SubmittedAt,
		StartedAt: base.StartedAt, FinishedAt: base.FinishedAt, Result: base.Result,
	}
	if base.Status == util.JobProcessing {
		if snap := getMapGenerateProgress(jobID); snap != nil {
			status.Progress = progressViewFromSnapshot(snap)
		} else if adtSnap, _ := d.Client.GetExportProgress(context.Background(), jobID); adtSnap != nil {
			percent := 0
			if adtSnap.TotalSteps > 0 {
				percent = minInt(100, (adtSnap.CompletedSteps*100)/adtSnap.TotalSteps)
			}
			var current *tileCoord
			if adtSnap.CurrentTile != nil {
				current = &tileCoord{X: adtSnap.CurrentTile.X, Y: adtSnap.CurrentTile.Y}
			}
			status.Progress = &mapGenerateProgressView{
				CompletedSteps: adtSnap.CompletedSteps, TotalSteps: adtSnap.TotalSteps, Phase: "adt",
				TileIndex: &adtSnap.TileIndex, TileCount: &adtSnap.TileCount, StepsPerTile: &adtSnap.StepsPerTile,
				CurrentTile: current, TaskName: adtSnap.TaskName, Percent: percent,
			}
		}
	} else if base.Status == util.JobDone && base.Result != nil {
		status.Progress = &mapGenerateProgressView{
			CompletedSteps: base.Result.TotalSteps, TotalSteps: base.Result.TotalSteps,
			Phase: "convert", TaskName: "Complete", Percent: 100,
		}
		clearMapGenerateProgress(jobID)
	} else if base.Status == util.JobFailed {
		clearMapGenerateProgress(jobID)
	}
	return status
}

func progressViewFromSnapshot(snap *mapGenerateSnapshot) *mapGenerateProgressView {
	return &mapGenerateProgressView{
		CompletedSteps: snap.CompletedSteps, TotalSteps: snap.TotalSteps, Phase: string(snap.Phase),
		TileIndex: snap.TileIndex, TileCount: snap.TileCount, StepsPerTile: snap.StepsPerTile,
		CurrentTile: snap.CurrentTile, TaskName: snap.TaskName,
		CreatureCompleted: snap.CreatureCompleted, CreatureTotal: snap.CreatureTotal,
		Percent: toProgressPercent(snap),
	}
}
